# Markov Decision Process


The lessons introduce the notion of Markov Decision Process (MDP) in three steps. This notion is import as a lot of RL problems can be formalised as MDPs

## Student Markov Chain


A state $S_t$ is Markov if and only if:

$\mathbb P [S_{t+1}|S_{t}]=\mathbb P [S_{t+1}|S_{1},...,S_{t}]$


A Markov Process is a tuple $(\mathcal S ,\mathcal P )$:
* $ \mathcal S $ a finite set of states
* $ \mathcal P $ a state transition probability matrix : $\mathcal P _{ss'}= \mathbb P [S_{t+1}=s'|S_{t}=s]$



A running example of the course on which I did some implementations is the student Markov chain
<p align="center">
	<img src="./Images/MP.png">
</p>
I implemented this markov chain in python be able to able probe the markov process:

In [1]:
import numpy as np
import matplotlib.pyplot as plt
from tqdm import tqdm

class StudentMarkovChain():
    """
    This class models the Markov process described in the lesson
    everything is hard coded, this class is not meant to be general
    """
    def __init__(self):
        #transition probabilities of each state
        self.transition = np.array([[0, 0.5 , 0 , 0 , 0 , 0.5 , 0 ],
                                    [0 , 0 , 0.8 , 0 , 0 , 0 , 0.2],
                                    [0 , 0 , 0 , 0.6 , 0.4 , 0 , 0],
                                    [0 , 0 , 0 , 0 , 0 , 0 , 1],
                                    [0.2 , 0.4 , 0.4 , 0 , 0 , 0 , 0],
                                    [0.1 , 0 , 0 , 0 , 0 , 0.9 , 0],
                                    [0 , 0 , 0 , 0 , 0 , 0 , 1]
                                    ])
        #name of states
        self.titles=["C1","C2","C3","Pass","Pub","FB","Sleep"]
        #first state
        self.state=0
        #the class will keep the history
        self.history = [self.titles[self.state]]

    #function to change state in the markov process
    def step(self):
        #Next state is picked following the probabilities of the transition matrix 
        self.state = np.random.choice(range(7),p=self.transition[self.state])
        self.history.append(self.titles[self.state])
        #if the state is  final
        if self.state != 6:
            #we return a state a bool telling if it is finished
            return self.state,False
        else:
            return self.state,True
    #function to restart
    def reboot(self):
        self.state = 0
        self.history = [self.titles[self.state]]


#function to run the markov chain
def main_markov():
    finished = False
    smc = StudentMarkovChain()
    while not finished:
        _,finished = smc.step()
    print(smc.history)


In [2]:
for i in range(10):
    main_markov()

['C1', 'C2', 'C3', 'Pub', 'C3', 'Pass', 'Sleep']
['C1', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'C1', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'C1', 'C2', 'C3', 'Pass', 'Sleep']
['C1', 'C2', 'C3', 'Pub', 'C3', 'Pub', 'C2', 'C3', 'Pass', 'Sleep']
['C1', 'C2', 'C3', 'Pass', 'Sleep']
['C1', 'FB', 'FB', 'C1', 'FB', 'FB', 'FB', 'C1', 'FB', 'FB', 'FB', 'C1', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'C1', 'FB', 'FB', 'FB', 'C1', 'C2', 'C3', 'Pass', 'Sleep']
['C1', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 'FB', 

You'll see after probing the markov process that students are enclined to pass an awful amount of time on facebook

## Markov Reward Process

A Markov Reward Process is a tuple $(\mathcal S ,\mathcal P,\mathcal R ,\gamma)$ where:
* $ \mathcal S $ a finite set of states
* $ \mathcal P $ a state transition probability matrix : $\mathcal P _{ss'}= \mathbb P [S_{t+1}=s'|S_{t}=s]$
* $ \mathcal R $ is a reward function, $ \mathcal R _s = \mathbb E [R_{t+1}|S_t = s] $ 
* $ \gamma $ is a discount factor, $\gamma \in [0,1]$

Here we present the Student reward process:

<p align="center">
	<img src="./Images/MRP.png">
</p>


In [3]:

class StudentMarkovRewardProcess(StudentMarkovChain):
    """
    Class to add rewards to the student markov chain
    it is inheriting the transition probabilities and the names from the markov chain
    """
    def __init__(self):
        """
        Constructor 
        """

        StudentMarkovChain.__init__(self)
        # we are adding here the rewards of the different states
        self.rewards=[-2,-2,-2,10,1,-1,0]
        #and the shape of the history includes the rewards 
        self.history[-1]=(self.history[-1],self.rewards[self.state])

    # change the step function of the markov chain to add the rewards    
    def step(self):
        state,finished = StudentMarkovChain.step(self)
        reward = self.rewards[state]
        self.history[-1]=(self.history[-1],reward)
        return self.state,finished,reward

    #function to restart
    def reboot(self):
        self.state = 0
        self.history = [(self.titles[self.state],self.rewards[self.state])]
        
#function to run the markov chain
def main_markov_reward():
    finished = False
    srp = StudentMarkovRewardProcess()
    while not finished:
        _,finished,_ = srp.step()
    print(srp.history)


In [4]:
#the history now has a reward attached to it
for i in range(10):
    main_markov_reward()

[('C1', -2), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('C1', -2), ('FB', -1), ('C1', -2), ('C2', -2), ('C3', -2), ('Pass', 10), ('Sleep', 0)]
[('C1', -2), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('C1', -2), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('C1', -2), ('C2', -2), ('C3', -2), ('Pass', 10), ('Sleep', 0)]
[('C1', -2), ('FB', -1), ('C1', -2), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('C1', -2), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB'

A central notion in reinforcement is the return $G_{t}$ that the agent can expect from now on to the end of the episode:
$$ G_{t}= R_{t+1}+\gamma R_{t+2}+ ... = \sum_{k=0}^{\infty} \gamma^{k} R_{t+k+1}$$

The state value is of a MRP is the expected return starting from state $s$:
$$ v(s) = \mathbb E [G_{t}|S_t = s]$$

In [5]:
def compute_return_markov_reward(verbose=True):
    gamma =0.9
    finished = False
    srp = StudentMarkovRewardProcess()
    while not finished:
        _,finished,_ = srp.step()
    
    rtn = sum([gamma**j*reward for j,(_,reward) in enumerate(srp.history)])
    if verbose:
        print(srp.history)
        print("Return of state 1: ",rtn)
    return rtn

In [6]:
compute_return_markov_reward()

[('C1', -2), ('C2', -2), ('C3', -2), ('Pub', 1), ('C1', -2), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('C1', -2), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('FB', -1), ('C1', -2), ('C2', -2), ('C3', -2), ('Pub', 1), ('C3', -2), ('Pass', 10), ('Sleep', 0)]
Return of state 1:  -11.7031921009334


-11.7031921009334

If we average the return obtained on the state 1 we will obtain the state value of state 1. it would just mean making a loop on the previous function and average the result

In [11]:
np.mean([compute_return_markov_reward(verbose=False) for _ in range(1000)])

-5.09440703804801

We have a value of -5.09 as the value of state 1 (the real value being -5 as we are about to see)

### Bellman equation for MRP
The bellman equation is the following one for MRP:
$$
v(s) = \mathbb E [R_{t+1} + \gamma v(S_{t+1})|S_{t}=s]
$$
As demonstrated in the lesson, using the bellman equation, we can derive an exact formula for the state values:


 $$\begin{align}
 \begin{bmatrix}
           v(1) \\
           \vdots \\
           v(n)
         \end{bmatrix}
         = 
         \begin{bmatrix}
           \mathcal R _1 \\
           \vdots \\
            \mathcal R _n
         \end{bmatrix} + \gamma 
         \begin{bmatrix}
           \mathcal P _{11} & \cdots & \mathcal P _{1n} \\
           \vdots & \cdots& \vdots\\
            \mathcal P _{n1} & \cdots & \mathcal P _{nn}
         \end{bmatrix}
         \begin{bmatrix}
           v(1) \\
           \vdots \\
           v(n)
         \end{bmatrix}
  \end{align}
  $$
  
  The solution of this system of equation (not doable for complex MRP) is given by:
  
  $$
  v = (I -\gamma \mathcal P)^{-1}\mathcal R
  $$

In [8]:
#computing the direct solution of the MDP 
#if gamma = 1 this computation is not doable
def direct_solution():
    smc = StudentMarkovRewardProcess()
    gamma = 0.9
    rev = np.dot(np.linalg.inv(np.eye(7)- gamma*smc.transition),smc.rewards)
    print(dict(zip(smc.titles,rev)))

direct_solution()

{'C1': -5.0127289100145225, 'C2': 0.9426552976939075, 'C3': 4.087021246797094, 'Pass': 10.0, 'Pub': 1.9083923522141464, 'FB': -7.637608431059512, 'Sleep': 0.0}


Here we find that the value of state 1 is -5 just like we computed just above, this computation has to be approximated in the case of a large MRP using
* Dynamic programming
* Monte Carlo
* TD learning

## Markov decision process

A Markov Decision Process is a tuple $(\mathcal S ,\mathcal A, \mathcal P,\mathcal R ,\gamma)$ (it is basically a reward process with decisions) where:
* $ \mathcal S $ a finite set of states
* $\mathcal A$ is a finite set of actions 
* $ \mathcal P $ a state transition probability matrix : $\mathcal P _{ss'}^{a}= \mathbb P [S_{t+1}=s'|S_{t}=s,A_t = a]$
* $ \mathcal R $ is a reward function, $ \mathcal R _s^a = \mathbb E [R_{t+1}|S_t = s,A_t = a] $ 
* $ \gamma $ is a discount factor, $\gamma \in [0,1]$

<p align="center">
	<img src="./Images/MDP.png">
</p>




In [9]:
#Here is a implementation of the student MDP that we will use in the next lessons
class MarkovStudentDecisionProcess():
    """
    Markov Decision Process of the Student graph presented in the lesson

    """
    def __init__(self):

        #less states than before 
        self.titles=["C1","C2","C3","FB","Sleep"]
        self.state=0
        #transition probabilites depend on the action taken by the agent
        # first line are the values for state one (c1), second line are the values for state 2 (c2),...
        #the transition matrix does have an extra dimension with the actions
        self.transition = [{"Study":[0,1,0,0,0],'Facebook':[0,0,0,1,0]},
                           {'Study':[0,0,1,0,0],'Sleep':[0,0,0,0,1]},
                           {'Study':[0,0,0,0,1],'Pub':[0.2,0.4,0.4,0,0]},
                           {'Facebook':[0,0,0,1,0],"Quit":[1,0,0,0,0]},
                           {"Sleep":[0,0,0,0,1]}
                            ]
        #the reward depends on the state and action taken by the agent
        self.reward=[{"Study":-2,'Facebook':-1},
                   {'Study':-2,'Sleep':0},
                   {'Study':10,'Pub':1},
                   {'Facebook':-1,"Quit":0},
                   {"Sleep":0}]


    # step depends now of the action taken by the agent    
    def step(self,action):
        reward = self.reward[self.state][action]
        self.state = np.random.choice(range(5),p=self.transition[self.state][action])
        
        
        finished=(action == "Sleep")
        return self.state,reward,finished
        #no history stored here the history will be stored by the agent
    def reboot(self):
        self.state = 0


class AgentRandom():
    def __init__(self):
        """
        Random agent has no will it just pick random actions
        """
        self.actions = [["Study",'Facebook'],
                       ['Study','Sleep'],
                       ['Study','Pub'],
                       ['Facebook',"Quit"],
                       ["Sleep"]]

    #selecting action randomly                   
    def select_action(self,state):
        return np.random.choice(self.actions[state])

    #virtual update
    def update(self,**kwargs):
        pass
    
# simple function to run the episode with a random agent
def run_episode():
    finished=False
    environement = MarkovStudentDecisionProcess()
    agent = AgentRandom()
    history = []
    while not finished:
        state = environement.state
        action = agent.select_action(state)
        _,reward,finished = environement.step(action)
        history.append((environement.titles[state],action,reward))
    print(history)



In [10]:
# running a sampled episode
run_episode()

[('C1', 'Study', -2), ('C2', 'Sleep', 0)]


you'll note that the history now includes a state, an action and a reward

I will not talk here of Partially observable Markov decision process or ergoticity sorry for those wanting it :-(
Next lesson is about dynamic programming